Add NGEN documentation (NETFramework-NGEN.md) (#9588)
dotnet/msbuild: documentation/NETFramework-NGEN.md

directly in the code, as it can be modified to work in non-standard environments just by editing the app config appropriately.

The resolver loading logic in MSBuild has been updated to call `Assembly.Load(AssemblyName)` where the `AssemblyName` specifies the
simple name of the assembly, e.g. `Microsoft.DotNet.MSBuildSdkResolver`, as well as its `CodeBase` (file path). This way the CLR assembly
loader will try to load the assembly into the default context first - a necessary condition for the native image to be used - and fall back
to LoadFrom if the simple name wasn't resolved.

### Microsoft.Build.NuGetSdkResolver

The NuGet resolver has many dependencies and its version is frequently changing. Due to the way Visual Studio is composed, MSBuild does
not know at its build time the exact version it will be loading at run-time. We would need a creative installer solution to be able to
have MSbuild.exe.config contain the right entries to load this resolver the same way as `Microsoft.DotNet.MSBuildSdkResolver``. A more
promising direction is loading this resolver into a separate AppDomain (see the section about NuGet.Frameworks below).

> The problem of JITting `Microsoft.Build.NuGetSdkResolver` remains unsolved for now.

Author: @jeffkl I believe you've also looked at optimizing NuGet resolver startup/JITting. Curious if you have any insights. Thank you!

Reviewer: I was going to investigate doing the following:

1. ILMerge everything into the NuGet-based MSBuild project SDK resolver so its only a single assembly
2. Stop changing the assembly version and only change the informational version
3. Ship it next to MSBuild by "inserting" it into MSBuild or add a `codeBase` entry in the `MSBuild.exe.config` since it can be relative.

#1 provided a surprising amount of improvement but some people are opposed to merging 
#2 should be a no-brainer in my opinion, I stopped rev'ing assembly versions a long time ago for everything I own

I did a test of this in early 2023 where I ILMerged the SDK resolver, manually edited the `MSBuild.exe.config` and built a local copy of MSBuild with changes to SDK resolver loading that worked against an assembly name rather than a path.  I saw a big performance increase for startup times when using the NuGet-based MSBuild project SDK resolver.  Let me know if you want me to dig up the branch where I made changes to MSBuild to load SDK resolvers by name instead of path.

Author: Thank you for the prompt response.

> Let me know if you want me to dig up the branch where I made changes to MSBuild to load SDK resolvers by name instead of path.

This is actually [already merged](https://github.com/dotnet/msbuild/pull/9439) and shipping in 17.9, affecting only `Microsoft.DotNet.MSBuildSdkResolver` which we were able to version-freeze. The impact on startup perf was indeed quite dramatic and it would be awesome if we can apply to the NuGet resolver as well.

Is there anything I can do to help make this happen?

Reviewer: The NuGet-based MSBuild project SDK resolver is loaded [via a manifest](https://github.com/dotnet/msbuild/blob/main/src/MSBuild/SdkResolvers/VS/Microsoft.Build.NuGetSdkResolver.xml), I had originally envisioned us updating the manifest reading code to look for a `name` element instead of a `path` but I suppose your change would work as well.  We'd need to ILMerge the SDK resolver and lock the version which is pretty simple.  Unfortunately, I'm not sure I'll have time in January to work on this.

## SDK resolvers

One class of assemblies loaded by MSBuild by path are SDK resolvers. MSBuild scans the `SdkResolvers` subdirectory to discover
the set of resolvers to use when evaluating projects. Extensible in theory, though in reality only a couple of resolvers actually
exist. Because resolvers ship as part of VS, it is not difficult to make sure their assemblies are properly NGENed. The hard part is
loading them with the regular `Assembly.Load` so the native images can be used. MSBuild cannot simply extend its probing path to
include the relevant subdirectories of `SdkResolvers` because 1) It is outside of the app directory cone for amd64 and arm64 versions
of `MSBuild.exe` and 2) Not all resolvers actually live under this directory; the system allows them to be placed anywhere.
It is unfortunately also not straightforward to add a binding redirect with a [`codeBase`](https://learn.microsoft.com/en-us/dotnet/framework/configure-apps/file-schema/runtime/codebase-element)
entry pointing to the right assemblies, because this requires knowing the exact assembly versions. 

### Microsoft.DotNet.MSBuildSdkResolver

> This is the base resolver, capable of resolving "in-box" SDKs that ship with the .NET SDK, and workloads. Since the resolver assembly

Reviewer 2: ```suggestion
This is the most-commonly-used resolver, capable of resolving "in-box" SDKs that ship with the .NET SDK and .NET SDK workloads. Since the resolver assembly
```

is located at a known path relative to MSBuild and has very few dependencies, none of which are used anywhere else, we have decided to

> freeze the version of the resolver plus dependencies, so that their full names can be specified in MSBuild.exe.config, e.g.

Reviewer 2: ```suggestion
freeze the version of the resolver plus dependencies, so that their full names can be specified in `MSBuild.exe.config`, e.g.
```

> ```

Reviewer 2: ````suggestion
```xml
````

    <dependentAssembly>
      <assemblyIdentity name="Microsoft.DotNet.MSBuildSdkResolver" culture="neutral" publicKeyToken="adb9793829ddae60" />
      <codeBase version="8.0.100.0" href=".\SdkResolvers\Microsoft.DotNet.MSBuildSdkResolver\Microsoft.DotNet.MSBuildSdkResolver.dll" />
    </dependentAssembly>
```

> Additionally, MSBuild.exe.config has the following entry, which enables us to refer to the resolver by simple name.

Reviewer 2: ```suggestion
Additionally, `MSBuild.exe.config` has the following entry, which enables us to refer to the resolver by simple name.
```

> ```

Reviewer 2: ````suggestion
```xml
````

<qualifyAssembly partialName="Microsoft.DotNet.MSBuildSdkResolver" fullName="Microsoft.DotNet.MSBuildSdkResolver, Version=8.0.100.0, Culture=neutral, PublicKeyToken=adb9793829ddae60" />
```

This has a small advantage compared to hardcoding `Microsoft.DotNet.MSBuildSdkResolver, Version=8.0.100.0, Culture=neutral, PublicKeyToken=adb9793829ddae60`
directly in the code, as it can be modified to work in non-standard environments just by editing the app config appropriately.

> The resolver loading logic in MSBuild has been updated to call `Assembly.Load(AssemblyName)` where the `AssemblyName` specifies the

Reviewer 2: ```suggestion
The resolver loading logic in MSBuild [has been updated](https://github.com/dotnet/msbuild/pull/9439) to call `Assembly.Load(AssemblyName)` where the `AssemblyName` specifies the
```

simple name of the assembly, e.g. `Microsoft.DotNet.MSBuildSdkResolver`, as well as its `CodeBase` (file path). This way the CLR assembly
loader will try to load the assembly into the default context first - a necessary condition for the native image to be used - and fall back
to LoadFrom if the simple name wasn't resolved.

### Microsoft.Build.NuGetSdkResolver

The NuGet resolver has many dependencies and its version is frequently changing, so the technique used for `Microsoft.DotNet.MSBuildSdkResolver`

> does not apply in its current state. However, the NuGet team it looking to address this by:

Reviewer 2: ```suggestion
does not apply in its current state. However, the NuGet team is looking to address this by:
```

> does not apply in its current state. However, the NuGet team it looking to address this by:

Reviewer 2: Got a link for this? I didn't see a bug on nuget/home beyond the long-closed https://github.com/NuGet/Home/issues/4350

Reviewer: Nothing specific yet, but the tracking epic is at https://github.com/NuGet/Home/issues/11441

Reviewer 2: That's plenty, thanks!

```suggestion
does not apply in its current state. However, the NuGet team is [looking to address this](https://github.com/NuGet/Home/issues/11441) by:
```

of cross-domain calls. The former is orders of magnitude smaller that the startup hit of JITting and the latter is negligible as long as
the types moved across the AppDomain boundary do not require expensive marshaling.

## Task assemblies

This is the proverbial elephant in the room. MSBuild learns about tasks dynamically as it parses project files. The `UsingTask`
element tends to specify the `AssemblyFile` attribute, pointing to the task assembly by path. Consequently MSBuild uses
`Assembly.LoadFrom` and no native images are loaded. Even task assemblies located in the SDK are problematic because MSBuild is
paired with an SDK on users machine at run-time. Unlike SDK resolvers and NuGet.Frameworks, which are part of the same installation
unit, this is a true dynamic inter-product dependency. Additionally, the task API is complex and involves a lot of functionality
provided to tasks via callbacks (e.g. logging) so the overhead of cross-domain calls may be significant. And that's assuming that
suitable native images exist in the first place, something that both VS and SDK installers would need to handle (task assemblies

> in each installed SDK are NGENed against each installed version of VS).

Reviewer 2: ```suggestion
in each installed SDK would need to be NGENed against each installed version of VS).
```
